Fix build with -DENABLE_GPU_PROCESS=ON in FontPlatformDataFreeType.cpp

psaavedra opened the pull request:
#### 231055e2de887f89b2856dd7af4c17a48e3dc337
<pre>
Fix build with -DENABLE_GPU_PROCESS=ON in FontPlatformDataFreeType.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=265037">https://bugs.webkit.org/show_bug.cgi?id=265037</a>

Reviewed by Carlos Garcia Campos.

Add the FontPlatformData::create() method for the freetype specialization
of the FontPlatformData factory.

The create() method brings back the lost code from the 4b35edbdccf87
&apos;[FreeType] Add initial implementation of variation fonts&apos; commit.

* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):

Canonical link: <a href="https://commits.webkit.org/271015@main">https://commits.webkit.org/271015@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f79db6fd8fc2ca0c7499b2c0c6cfe0060b9e088d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27077 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5693 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28313 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29285 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24758 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27530 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3088 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24606 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4511 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23241 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3940 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4032 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24238 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29921 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24714 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24649 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30228 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2237 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28144 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5509 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6508 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4510 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4422 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
```diff
@@ -322,4 +322,40 @@ HbUniquePtr<hb_font_t> FontPlatformData::createOpenTypeMathHarfBuzzFont() const
 }
 #endif
 
+#if PLATFORM(WPE)
```
Why is this WPE specific?

psaavedra replied:
Oh, this is a mistake. Since this file is freetype specific this conditional check is not necessary for sure.

carlosgcampos commented on Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
```diff
@@ -322,4 +322,40 @@ HbUniquePtr<hb_font_t> FontPlatformData::createOpenTypeMathHarfBuzzFont() const
+    if (FcPatternGetInteger(pattern, FC_SPACING, 0, &spacing) == FcResultMatch && spacing == FC_MONO)
+        fixedWidth = true;
+
+    return FontPlatformData(fontFace.get(), adoptRef(pattern), data.m_size, fixedWidth, data.m_syntheticBold, data.m_syntheticOblique, data.m_orientation, custom);
```
Shouldn't we use custom->m_fontFace if custom is not nullptr?

carlosgcampos commented on Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
```diff
@@ -322,4 +322,40 @@ HbUniquePtr<hb_font_t> FontPlatformData::createOpenTypeMathHarfBuzzFont() const
+{
+    Attributes result(m_size, m_orientation, m_widthVariant, m_textRenderingMode, m_syntheticBold, m_syntheticOblique);
+
+    return result;
```
I guess this could be return Attributes(m_size, m_orientation, m_widthVariant, m_textRenderingMode, m_syntheticBold, m_syntheticOblique);

carlosgcampos commented on Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
```diff
@@ -322,4 +322,40 @@ HbUniquePtr<hb_font_t> FontPlatformData::createOpenTypeMathHarfBuzzFont() const
+    if (custom->m_fontFace)
+        fontFace = custom->m_fontFace;
+    else
+        fontFace = adoptRef(cairo_ft_font_face_create_for_pattern(pattern));
```
custom can be nullptr, here. Also we don't need the pattern if we already hace a font face from custom.

psaavedra replied:
Good point thanks for the hint. Added to the latest version of the patch.

carlosgcampos commented on Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
```diff
@@ -322,4 +322,41 @@ HbUniquePtr<hb_font_t> FontPlatformData::createOpenTypeMathHarfBuzzFont() const
+    // so Fontconfig will not affect the hint style, but it may disable hinting completely.
+    static std::once_flag flag;
+    static FcPattern* pattern = nullptr;
+    static bool fixedWidth = false;
```
Move this to the else branch too

psaavedra replied:
I don't know if this is necessary. The  `pattern` and the `fixedWidth`  static definitions are used in L354. Because they are static definitions inside a function then the data is allocated statically, visible only inside the function  but, more important, only [initialized the first time the block is entered](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables).

Also if we let this as it is code right now we make more explicit that those variables will be always initialized to a default value, no matters of what part of the if-then-else is chosen.

carlosgcampos replied:
It's not necessary, it just makes explicit that they are only used there and the comment only applies to the else branch too.

psaavedra replied:
this is true for the comment and also for the `flag` variable but still we need to define the fixedWidth and the pattern variables in the main block of the function to be available for the return statement (what uses them). I will update the patch.

psaavedra committed:
Fix build with -DENABLE_GPU_PROCESS=ON in FontPlatformDataFreeType.cpp
https://bugs.webkit.org/show_bug.cgi?id=265037

Reviewed by Carlos Garcia Campos.

Add the FontPlatformData::create() method for the freetype specialization
of the FontPlatformData factory.

The create() method brings back the lost code from the 4b35edbdccf87
'[FreeType] Add initial implementation of variation fonts' commit.

* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):

Canonical link: https://commits.webkit.org/271015@main
